opm/io/eclipse/OutputStream.hpp

```diff
@@ -0,0 +1,241 @@
+/*
+  Copyright (c) 2019 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_IO_OUTPUTSTREAM_HPP_INCLUDED
+#define OPM_IO_OUTPUTSTREAM_HPP_INCLUDED
+
+#include <ios>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace Opm { namespace ecl {
+
+    class EclOutput;
+
+}} // namespace Opm::ecl
+
+namespace Opm { namespace ecl { namespace OutputStream {
+
+    struct Formatted { bool set; };
+    struct Unified   { bool set; };
+
+    /// Abstract representation of an ECLIPSE-style result set.
+    struct ResultSet
+    {
+        /// Output directory.  Commonly "." or location of run's .DATA file.
+        std::string outputDir;
+
+        /// Base name of simulation run.
+        std::string baseName;
+    };
+
+    /// File manager for restart output streams.
+    class Restart
+    {
+    public:
+        /// Constructor.
+        ///
+        /// \param[in] rset Output directory and base name of output stream.
+        ///
+        /// \param[in] fmt Whether or not to create formatted output files.
+        ///
+        /// \param[in] unif Whether or not to create unified output files.
+        explicit Restart(ResultSet        rset,
+                         const Formatted& fmt,
+                         const Unified&   unif);
+
+        ~Restart();
+
+        Restart(const Restart& rhs) = delete;
+        Restart(Restart&& rhs);
+
+        Restart& operator=(const Restart& rhs) = delete;
+        Restart& operator=(Restart&& rhs);
+
+        /// Opens file stream pertaining to restart of particular report
+        /// step and also outputs a SEQNUM record in the case of a unified
+        /// output stream.
+        ///
+        /// Must be called before accessing the stream object through the
+        /// stream() member function.
+        ///
+        /// \param[in] seqnum Sequence number of new report.  One-based
+        ///    report step ID.
+        void prepareStep(const int seqnum);
+
+        const ResultSet& resultSetDescriptor() const
+        {
+            return this->rset_;
+        }
+
+        bool formatted() const
+        {
+            return this->formatted_;
+        }
+
+        bool unified() const
```

Comment:
Do you really need to internalize these bools - I have guess that they should already be in the `EclOutput`class and that should be enough?

Reply:
> Do you really need to internalize these bools

Maybe[tm].  It makes unit-testing the `outputFileName()` function overloads easier, although the filename function for `Restart`s could arguably be a private member function.  Also, unless we want to do all file operations in the constructor of `Restart`&mdash;which in turn would make keeping the output file open across all timesteps more difficult&mdash;then we really do need a separate copy of the flags.

Reply:
> It makes unit-testing the outputFileName() function overloads easier, 

Now - I am all for testing; but I have changed my ways slightly - and of the mental changes I have made is that if the ability to test is affecting the details of your API you should at least twice - my experience is that you are then testing at the wrong level.

> Also, unless we want to do all file operations in the constructor of Restart—which in turn would make keeping the output file open across all timesteps more difficult—then we really do need a separate copy of the flags.

If, then and when. Right now&trade; we do not need - then we remove them.